### PR TITLE
[12.0][ADD] maintenance_request_time_performance

### PR DESCRIPTION
**Pull request by lmiguens-solvos**

### CODE REVIEW

File: maintenance_request_time_performance/data/maintenance_data.xml [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+<odoo>
+<data>
+
+    <!-- Standard stages: default in progress stage -->
+    <function
+            id="function_default_stage_in_progress"
+            model="maintenance.stage"
+            name="update_default_stage_in_progress"/>
+
+</data>
```
Comment (dalonsod) [inline maintenance_request_time_performance/data/maintenance_data.xml]: I think this can be achieved in a better way (locally tested):

```suggestion
    <data noupdate="1">
        <record id="maintenance.stage_1" model="maintenance.stage">
            <field name="in_progress" eval="True" />
        </record>
    </data>
```

File: maintenance_request_time_performance/models/maintenance.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+    in_progress = fields.Boolean(
+        'In Progress',
+        default=False)
+
+    @api.model
+    def update_default_stage_in_progress(self):
+        stage = self.env.ref('maintenance.stage_1')
+        if stage:
+            stage.write({'in_progress': True})
```
Comment (dalonsod) [inline maintenance_request_time_performance/models/maintenance.py]: This method is no longer needed

File: maintenance_request_time_performance/models/maintenance.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+    start_task_datetime = fields.Datetime(
+        string='Start Date and time of maintenance task completion')
+    end_task_datetime = fields.Datetime(
+        string='End Date and time of maintenance task completion')
+    real_task_duration = fields.Float(
+        compute='_compute_real_task_duration',
+        string='Real Duration',
+        store=True)
+
```
Comment (dalonsod) [inline maintenance_request_time_performance/models/maintenance.py:75]: a computed, stored field name `duration_difference` can be added and used in pivot view

File: maintenance_request_time_performance/models/maintenance.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+                if request.stage_id.done and not request.start_task_datetime:
+                    raise ValidationError(
+                        _("It can´t be done without Start task completion"))
+                # if stage_id changes to in_progress -> fill start_task_datetime
+                if request.stage_id.in_progress \
+                        and not request.start_task_datetime:
+                    request.start_task_datetime = fields.Datetime.now()
+                # if stage_id changes to done stage -> fill end_task_datetime
+                if request.stage_id.done and not request.end_task_datetime:
+                    request.end_task_datetime = fields.Datetime.now()
```
Comment (dalonsod) [inline maintenance_request_time_performance/models/maintenance.py]: Well, I finally agree with you, setting a field within an `@api.constrains` method is a bad procedure.

In other hand, I've tested start and end datetime fields and I found these errors:
- It's possible to change start datetime and set it after end datetime.
- It's possible to unset start datetime when end time is already filled.

So I make the following proposal:
- change `start_task_datetime` and `end_task_datetime` definition and make them with a compute method, `store=True` and `readonly=False`, so they can be computed but also editable.
- the compute method (e.g. `_compute_task_datetimes`) can be used, when `stage_id` is changed, to update that values when empty (this part can be recovered from `check_maintenance_request_stage` current method).
- define another method (e.g. `check_maintenance_request_task_datetimes`), when any of the task datetime fields is manually changed, to only cover the validation part of the `check_maintenance_request_stage` current method

File: maintenance_request_time_performance/views/maintenance_views.xml [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+    <!-- New search view: MaintenanceRequest - added start and end datetime -->
+    <record id="hr_equipment_request_view_search" model="ir.ui.view">
+        <field name="name">equipment.request.search</field>
+        <field name="model">maintenance.request</field>
+        <field name="inherit_id" ref="maintenance.hr_equipment_request_view_search"/>
+        <field name="sequence" eval="18"/>
+        <field name="arch" type="xml">
+            <xpath expr="//filter[@name='progress']" position="after">
+                <filter name="progresstoday" string="Start Today"
```
Comment (dalonsod) [inline maintenance_request_time_performance/views/maintenance_views.xml]: Typo

```suggestion
                <filter name="progresstoday" string="Started Today"
```

**Review by dalonsod — CHANGES_REQUESTED** [review]
See comments. Anyway, great job, please be patient with my change requests :(

**Comment by dalonsod** [comment]
@lmiguens-solvos I've made some changes since your recent commit:
- Making `start_request_datetime` and `end_request_datetime` computed and stored was also a bad idea, sorry for my previous requirements! It only works if a field is completely computed, but in our case we sometimes need to preserve the value (e.g. if it is manually stored and it's correct) => stored values, if computed, are completely unset and old value can't be recovered! So, I made them as regular fields and switched from `@api.depends` to `@api.onchange` related to `stage_id` value.
- Added some stuff in order to make (un)required those values on form editing, depending on the current stage. Also real duration is only shown when both values are filled.
- Added a temporary workaround for module installation and default datetime values.

Could you test these new changes?

**Comment by dalonsod** [comment]
@lmiguens-solvos test your recent changes, look 👍 

Anyway, I've made another implementation of the `post_init_hook` method: https://github.com/solvosci/slv-maintenance/blob/99f68ee12dece80ba1ebd6998a17f10810b834a4/maintenance_request_time_performance/hooks.py#L54

Could you take a look? I think it keeps functionality and the code is quite simpler, but I've not tested it yet. If didn't work, feel free for adapt it...

File: maintenance_request_time_performance/hooks.py [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+            ('stage_id.in_progress', '=', True)])
+        rec_done = MaintenanceRequest.search([
+            ('stage_id.done', '=', True)])
+        for req in (rec_in_progress | rec_done):
+            req.write({
+                'start_request_datetime':
+                    (req.schedule_date
+                     if req.schedule_date != req.create_date.date() else None)
+                    or (req.request_date
+                        if req.request_date != req.create_date.date() else None)
```
Comment (dalonsod) [inline maintenance_request_time_performance/hooks.py]: Line with 80 chars, it should have< 80

File: maintenance_request_time_performance/hooks.py [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+                    or req.create_date,
+                'end_request_datetime':
+                    req.stage_id.done
+                    and ((req.close_date
+                          if req.close_date != req.write_date.date() else None)
+                         or req.write_date),
+            })
```
Comment (dalonsod) [inline maintenance_request_time_performance/hooks.py]: No newline at the end of the file

**Review by dalonsod — CHANGES_REQUESTED** [review]
Fix commented issues, squash commits and PR will be ready for merge!!

File: maintenance_request_time_performance/models/maintenance.py [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+        compute='_compute_real_request_duration',
+        string='Real Duration',
+        store=True)
+    duration_difference = fields.Float(
+        compute='_compute_duration_difference',
+        string='Difference Duration',
+        store=True)
+    completion_delay = fields.Float(
+        compute='_compute_completion_delay',
+        string='Completion Delay',
```
Comment (dalonsod) [inline maintenance_request_time_performance/models/maintenance.py]: You can remove this line, Odoo will set `string` value to "Completion Delay" anyway

File: maintenance_request_time_performance/models/maintenance.py [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+        for req in self:
+            req.start_request_datetime_required = \
+                req.stage_id.done or req.stage_id.in_progress
+
+    @api.depends('end_request_datetime')
+    def _compute_end_request_date(self):
+        for req in self:
+            req.end_request_date = \
+                (req.end_request_datetime.date()
+                 if req.end_request_datetime else None)
```
Comment (dalonsod) [inline maintenance_request_time_performance/models/maintenance.py]: Alternative and clearer code:

```suggestion
            req.end_request_date = \ 
                req.end_request_datetime and req.end_request_datetime.date()
```

File: maintenance_request_time_performance/models/maintenance.py [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+
+    @api.depends('duration', 'real_request_duration')
+    def _compute_duration_difference(self):
+        for req in self:
+            difference = 0
+            if req.duration and req.real_request_duration:
+                difference = req.real_request_duration - req.duration
+            req.duration_difference = difference
+
+    @api.depends('maintenance_type', 'schedule_date', 'duration', 'start_request_datetime', 'end_request_datetime')
```
Comment (dalonsod) [inline maintenance_request_time_performance/models/maintenance.py]: Line with >= 80 chars

File: maintenance_request_time_performance/views/maintenance_views.xml [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+                <field name="end_request_datetime"
+                       attrs="{'required': [('end_request_datetime_required', '=', True)]}"/>
+                <label for="real_request_duration"/>
+                <div attrs="{'invisible': ['|', ('start_request_datetime', '=', False), ('end_request_datetime', '=', False)]}">
+                    <field name="real_request_duration" widget="float_time" class="oe_inline"/> <span class="ml8">hours</span>
+                </div>
+                <label for="completion_delay"/>
+                <div attrs="{'invisible': ['|', ('schedule_date', '=', False), ('end_request_datetime', '=', False)]}">
+                    <field name="completion_delay" widget="float_time" class="oe_inline"/> <span class="ml8">hours</span>
+                </div>
```
Comment (dalonsod) [inline maintenance_request_time_performance/views/maintenance_views.xml]: Should `completion_delay` be visible for everyone in the `form` view, or not? I'm not sure...

**Review by dalonsod — CHANGES_REQUESTED** [review]
Minor changes & doubts. When attended, I'll test them with a large database

The most recent improvement (`completion_delay`) looks 👍

File: maintenance_request_time_performance/models/maintenance.py (hunk 8 above)
Comment (dalonsod) [reply maintenance_request_time_performance/models/maintenance.py]: Pending change

File: maintenance_request_time_performance/models/maintenance.py (hunk 9 above)
Comment (dalonsod) [reply maintenance_request_time_performance/models/maintenance.py]: Pending change (only suggestion)

File: maintenance_request_time_performance/models/maintenance.py [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+                 'duration',
+                 'start_request_datetime',
+                 'end_request_datetime')
+    def _compute_completion_delay(self):
+        for req in self:
+            diff = None
+            if req.end_request_datetime:
+                schedule_dt = (req.schedule_date
+                               or req.start_request_datetime
+                               or req.create_date)
```
Comment (dalonsod) [inline maintenance_request_time_performance/models/maintenance.py]: if `req.end_request_datetime` is filled, `req.start_request_datetime` is always filled too, isn't it? If I'm right, this line is never evaluated

File: maintenance_request_time_performance/models/maintenance.py [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+                        and req.start_request_datetime:
+                    # if has no time we put the start_request time
+                    schedule_dt = schedule_dt.replace(
+                        hour=req.start_request_datetime.hour,
+                        minute=req.start_request_datetime.minute,
+                        second=req.start_request_datetime.second)
+                if req.duration:
+                    # add theoretical duration to schedule_date
+                    schedule_dt = schedule_dt + timedelta(hours=req.duration)
+                diff = req.end_request_datetime - schedule_dt
```
Comment (dalonsod) [inline maintenance_request_time_performance/models/maintenance.py:116]: I think the code below is equivalent (`req.duration` equals `0` if not filled), bit I didn't check it:

```suggestion
                diff = req.end_request_datetime - (schedule_dt + timedelta(hours=req.duration))
```

File: maintenance_request_time_performance/models/maintenance.py [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+            delay = None
+            if diff:
+                delay = diff.total_seconds() / 3600
+            req.completion_delay = delay
```
Comment (dalonsod) [inline maintenance_request_time_performance/models/maintenance.py]: Alternative code (only suggestion, not checked):

```suggestion
            req.completion_delay = diff and (diff.total_seconds() / 3600)
```

File: maintenance_request_time_performance/hooks.py [hunk 15]
```diff
@@ -0,0 +1,44 @@
+from odoo import _, api, SUPERUSER_ID
+from datetime import datetime, timedelta
+
+
+def post_init_hook(cr, registry):
+    with api.Environment.manage():
+        env = api.Environment(cr, SUPERUSER_ID, {})
+        maintenance_request = env['maintenance.request']
```
Comment (dalonsod) [inline maintenance_request_time_performance/hooks.py]: It seems there's an Odoo/OCA convention with model object naming: if the model is `maintenance.request`, an object for the model itself (not for a record set) should be named `MaintenanceRequest`. So, I'd revert this change...

**Review by dalonsod — CHANGES_REQUESTED** [review]
Only cosmetic and better code practices comments

File: maintenance_request_time_performance/models/maintenance.py (hunk 8 above)
Comment (lmiguens-solvos) [reply maintenance_request_time_performance/models/maintenance.py]: ok. This is because I put in "string" exactly the name of the field, right?

File: maintenance_request_time_performance/models/maintenance.py (hunk 9 above)
Comment (lmiguens-solvos) [reply maintenance_request_time_performance/models/maintenance.py]: ok. The same on line 56...

File: maintenance_request_time_performance/views/maintenance_views.xml (hunk 11 above)
Comment (lmiguens-solvos) [reply maintenance_request_time_performance/views/maintenance_views.xml]: Completion_delay is visible for everyone according to user permissions, but it is invisible when (now) end_request_datetime is False

File: maintenance_request_time_performance/models/maintenance.py (hunk 12 above)
Comment (lmiguens-solvos) [reply maintenance_request_time_performance/models/maintenance.py]: You are right. The last element of the "or" does not make sense

File: maintenance_request_time_performance/models/maintenance.py (hunk 13 above)
Comment (lmiguens-solvos) [reply maintenance_request_time_performance/models/maintenance.py:116]: The difference is if req.duration is null...

File: maintenance_request_time_performance/models/maintenance.py (hunk 14 above)
Comment (lmiguens-solvos) [reply maintenance_request_time_performance/models/maintenance.py]: ok. I like it. I'll check it

File: maintenance_request_time_performance/hooks.py (hunk 15 above)
Comment (lmiguens-solvos) [reply maintenance_request_time_performance/hooks.py]: ok. I changed because Pycharm was giving me a warning "Variable in function shoud be lowercase". I'll revert this change.

File: maintenance_request_time_performance/views/maintenance_views.xml (hunk 11 above)
Comment (lmiguens-solvos) [reply maintenance_request_time_performance/views/maintenance_views.xml]: I am unable to check end_request_datetime is False. I changed real_request_duration invisible attribute too. Both with invisible if real_request_duration=0. I think it's a solution, but I don´t know if it´s the best solution.

File: maintenance_request_time_performance/models/maintenance.py (hunk 8 above)
Comment (dalonsod) [reply maintenance_request_time_performance/models/maintenance.py]: That's right, the default `string` attribute for `some_custom_data` field is "Some Custom Data"

File: maintenance_request_time_performance/views/maintenance_views.xml (hunk 11 above)
Comment (dalonsod) [reply maintenance_request_time_performance/views/maintenance_views.xml]: > Completion_delay is visible for everyone according to user permissions, but it is invisible when (now) end_request_datetime is False

My question is philosophical. Since `completion_delay` could show a _negative_ data for the technician, I'm not sure if this field should only be visible for a manager role... Anyway, in the `pivot` view is always visible too...



> I am unable to check end_request_datetime is False. I changed real_request_duration invisible attribute too. Both with invisible if real_request_duration=0. I think it's a solution, but I don´t know if it´s the best solution.

OK, I'll take a look for that unexpected behavior.

File: maintenance_request_time_performance/models/maintenance.py (hunk 13 above)
Comment (dalonsod) [reply maintenance_request_time_performance/models/maintenance.py:116]: OK, good to know!

**Comment by dalonsod** [comment]
Reviewed and some changes made:
- Removed some unnecessary comments.
- Simplified `string` definition for `start_request_datetime` and `end_request_datetime`. The original one is moved to `help` attribute.
- Added some of the new fields to `tree` view.
- Restored `end_request_datetime` to the real duration and completion delay invisible condition. I tested it and worked correctly for me.
- Add test coverage. @lmiguens-solvos you can take a look and run it with `-u maintenance_request_time_performace --test-enable` and looking console output.

If changes are OK, I'll test the addon with a large database prior to merge the PR

**Comment by dalonsod** [comment]
Since last commit,
- Improved test expressions.
- Deleted unknown attribute `sequence` for search view.
- Tested on a large database.
- Squashed commits

Ready to merge!

**Review by dalonsod — APPROVED** [review]
